node/primitives/src/lib.rs

```diff
@@ -405,3 +412,227 @@ impl SignedDisputeStatement {
+			DisputeStatement::Valid(v) => v,
+			_ => {
+				tracing::debug!("Valid statement had no valid kind.");
+				return None
+			}
+		};
+
+		let invalid_kind = match invalid_statement.statement() {
+			DisputeStatement::Invalid(v) => v,
+			_ => {
+				tracing::debug!("Invalid statement had no invalid kind.");
+				return None
+			}
+		};
+
+		let valid_vote = ValidDisputeVote {
+			validator_index: valid_index,
+			signature: valid_statement.validator_signature().clone(),
+			kind: valid_kind.clone(),
+		};
+
+		let invalid_vote = InvalidDisputeVote {
+			validator_index: invalid_index,
+			signature: invalid_statement.validator_signature().clone(),
+			kind: invalid_kind.clone(),
+		};
+		Some(DisputeMessage(UncheckedDisputeMessage {
+			candidate_receipt,
+			session_index,
+			valid_vote,
+			invalid_vote,
+		}))
+	}
+
+	/// Read only access to the candidate receipt.
+	pub fn candidate_receipt(&self) -> &CandidateReceipt {
+		&self.0.candidate_receipt
+	}
+
+	/// Read only access to the `SessionIndex`.
+	pub fn session_index(&self) -> SessionIndex {
+		self.0.session_index
+	}
+
+	/// Read only access to the invalid vote.
+	pub fn invalid_vote(&self) -> &InvalidDisputeVote {
+		&self.0.invalid_vote
+	}
+
+	/// Read only access to the valid vote.
+	pub fn valid_vote(&self) -> &ValidDisputeVote {
+		&self.0.valid_vote
+	}
+}
+
+impl UncheckedDisputeMessage {
+	/// Try to recover the two signed dispute votes from an UncheckedDisputeMessage.
+	pub fn try_into_signed_votes(self, session_info: &SessionInfo)
+		-> Result<(CandidateReceipt, (SignedDisputeStatement, ValidatorIndex), (SignedDisputeStatement, ValidatorIndex)), ()>
+	{
+		let Self {
+			candidate_receipt,
+			session_index,
+			valid_vote,
+			invalid_vote,
+		} = self;
+		let candidate_hash = candidate_receipt.hash();
+
+		let vote_valid = {
+			let ValidDisputeVote {
+				validator_index,
+				signature,
+				kind,
+			} = valid_vote;
+			let validator_public = session_info.validators.get(validator_index.0 as usize).ok_or(())?.clone();
+
+			(
+				SignedDisputeStatement::new_checked(
+					DisputeStatement::Valid(kind), candidate_hash, session_index, validator_public, signature
+				)?,
+				validator_index
+			)
+		};
+
+		let vote_invalid = {
+			let InvalidDisputeVote {
+				validator_index,
+				signature,
+				kind,
+			} = invalid_vote;
+			let validator_public = session_info.validators.get(validator_index.0 as usize).ok_or(())?.clone();
+
+			(
+				SignedDisputeStatement::new_checked(
+					DisputeStatement::Invalid(kind), candidate_hash, session_index, validator_public, signature
+				)?,
+				validator_index
+			)
+		};
+
+		Ok((candidate_receipt, vote_valid, vote_invalid))
+	}
+}
+
+impl From<DisputeMessage> for UncheckedDisputeMessage {
+	fn from(message: DisputeMessage) -> Self {
+		message.0
+	}
+}
+
+/// Any invalid vote (currently only explicit).
+#[derive(Clone, Encode, Decode, Debug)]
+pub struct InvalidDisputeVote {
```

Comment:
I think that the `DisputeStatementSet` definition could be improved by the use of this and `ValidDisputeVote` as opposed to the tuples we are currently using. We'd also gain consistency by that change.

Reply:
Not really [`DisputeStatementSet`](https://github.com/paritytech/polkadot/blob/a620156c0cdb46991b8eae89b99d1941aa8d9e18/primitives/src/v1/mod.rs#L1182) does not differentiate between valid and invalid votes.